### PR TITLE
[CALCITE-2631] + [CALCITE-2639]

### PR DESCRIPTION
**Pull request by jcamachor**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 1]
```diff
@@ -676,52 +676,64 @@ private RexNode simplifyCoalesce(RexCall call) {
...
+    RelDataType caseType = call.getType();
 
+    boolean merged = false;
+    CaseBranch lastBranch = null;
     List<CaseBranch> branches = new ArrayList<>();
     for (CaseBranch branch : inputBranches) {
       // simplify the condition
-      RexNode newCond = branchSimplifier.simplify(branch.cond, RexUnknownAs.FALSE);
+      RexNode newCond = condSimplifier.simplify(branch.cond, RexUnknownAs.FALSE);
+      if (newCond.isAlwaysFalse()) {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Should this be placed after `simplifyAndAddToBranches` ?

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 2]
```diff
@@ -676,52 +676,64 @@ private RexNode simplifyCoalesce(RexCall call) {
...
+      RexNode newValue = valueSimplifier.simplify(branch.value, unknownAs);
 
-      branches.add(new CaseBranch(newCond, newValue));
-    }
-
-    // remove branches with invalid conditions
-    branches.removeIf(branch -> branch.cond.isAlwaysFalse());
+      // create new branch
+      if (lastBranch != null) {
+        if (lastBranch.value.toString().equals(newValue.toString())
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: I guess this never works since `newCond` is overwritten on each `for` iteration

Comment (jcamachor) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: newCond becomes part of lastBranch, not sure if you are referring to that.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 3]
```diff
@@ -738,30 +750,22 @@ private RexNode simplifyCase(RexCall call, RexUnknownAs unknownAs) {
...
+   * If boolean is true, first simplify when clause in branch, then add new case to branches.
+   * Otherwise, simply add input case to branches.
+   */
+  private void simplifyAndAddToBranches(boolean simplifyCond, RexSimplify simplifier,
+      CaseBranch branch, List<CaseBranch> branches) {
+    if (simplifyCond) {
+      // the previous branch was merged, time to simplify it and
+      // add it to the final result
+      branch = new CaseBranch(
+          simplifier.simplify(branch.cond, RexUnknownAs.FALSE), branch.value);
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Pushed a new commit for this one.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java (hunk 2 above)
Comment (vlsi) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: That's truly hard to follow

Comment (jcamachor) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: This lets us simplifying the merged condition only once. I will add some additional comments with explanation so it becomes a bit easier to know what is going on.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 4]
```diff
@@ -676,52 +676,72 @@ private RexNode simplifyCoalesce(RexCall call) {
...
+          // hence we create a new composite condition and we do not add it to
+          // the final branches for the time being
+          newCond = rexBuilder.makeCall(SqlStdOperatorTable.OR, lastBranch.cond, newCond);
+          merged = true;
+        } else {
+          // if we reach here, the new branch is not mergeable with the last one,
+          // hence we are going to add the last branch to the final branches.
+          // if the last branch was merged, then we will simplify it first.
+          // otherwise, we just add it
+          branches.add(generateBranch(merged, condSimplifier, lastBranch));
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: You probably want to check isAlways for the `simplified` variation version of the branch (and to `generateBranch` below as well)

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java [hunk 5]
```diff
@@ -676,52 +676,72 @@ private RexNode simplifyCoalesce(RexCall call) {
...
+        CaseBranch.fromCaseOperands(rexBuilder, new ArrayList<>(call.getOperands()));
 
     // run simplification on all operands
-    RexSimplify branchSimplifier = this;
-    RelDataType branchType = call.getType();
+    RexSimplify condSimplifier = this.withPredicates(RelOptPredicateList.EMPTY);
+    RexSimplify valueSimplifier = this;
+    RelDataType caseType = call.getType();
 
+    boolean merged = false;
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: In my point of view `conditionNeedsSimplify` would make more sense

Comment (jcamachor) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Makes sense, I changed it.

File: core/src/main/java/org/apache/calcite/rex/RexSimplify.java (hunk 4 above)
Comment (jcamachor) [reply core/src/main/java/org/apache/calcite/rex/RexSimplify.java]: Good point, I missed that one. I have changed it in latest commit.

**Comment by jcamachor** [comment]
@vlsi , thanks for reviewing. Please let me know if latest version looks good to you so I can merge it.

**Comment by vlsi** [comment]
@jcamachor , this looks good to me modulo tests.
e.g. the tests for `two branches collapse to "always true"`:
`case when x is true then 1 when x is not true then 1 else 2 end`
`case when x is true then 1 when true then 1 else 2 end`

**Comment by jcamachor** [comment]
@vlsi , just pushed those tests in last commit and simplification LGTM. Can you double check? If they are fine, I will merge this PR. Thanks

File: core/src/test/java/org/apache/calcite/test/RexProgramTest.java [hunk 6]
```diff
@@ -1739,6 +1739,41 @@ private void checkExponentialCnf(int n) {
...
+    RexNode literal2 = literal(2);
+
+    RexNode caseNode = case_(firstCond, literal1, secondCond, literal1, literal2);
+
+    RexNode result =
+        simplify.simplifyUnknownAs(caseNode, RexUnknownAs.FALSE);
+    assertThat(result.getType().isNullable(), is(false));
+    assertThat(result.getType().getSqlTypeName(), is(SqlTypeName.INTEGER));
+    RexNode expectedResult = case_(or(x, not(x)), literal1, literal2);
+    assertThat(result.toString(), is(expectedResult.toString()));
```
[the remark this answers is not in the document]
Comment (jcamachor) [reply core/src/test/java/org/apache/calcite/test/RexProgramTest.java]: I did not know we had that utility method, makes sense, I have changed it accordingly.

**Comment by vlsi** [comment]
LGTM
